src/core/thread/wakeup_tx_scheduler.hpp

```diff
@@ -0,0 +1,125 @@
+#include "common/non_copyable.hpp"
+#include "common/timer.hpp"
+#include "mac/mac.hpp"
+
+namespace ot {
+
+class Child;
+
+/**
+ * This class implements wake-up sequence TX scheduling functionality.
+ *
+ */
+class WakeupTxScheduler : public InstanceLocator, private NonCopyable
+{
+    friend class Mac::Mac;
+
+public:
+    /**
+     * Initializes the wake-up sequence TX scheduler object.
+     *
+     * @param[in]  aInstance   A reference to the OpenThread instance.
+     *
+     */
+    explicit WakeupTxScheduler(Instance &aInstance);
+
+    /**
+     * Initiates the wake-up sequence to a Wake-up End Device.
+     *
+     * @param[in] aWedAddress The extended address of the Wake-up End Device.
+     * @param[in] aIntervalUs An interval between consecutive wake-up frames (in microseconds).
+     * @param[in] aDurationMs Duration of the wake-up sequence (in milliseconds).
+     *
+     * @retval kErrorNone         Successfully started the wake-up sequence.
+     * @retval kErrorInvalidState This or another device is currently being woken-up.
+     */
+    Error WakeUp(const Mac::ExtAddress &aWedAddress, uint16_t aIntervalUs, uint16_t aDurationMs);
+
+    /**
+     * Returns the connection window used by this device.
+     *
+     * The connection window is amount of time that this device waits for an initial link establishment message after
+     * sending the last wake-up frame.
+     *
+     * @returns Connection window in the units of microseconds.
+     */
+    uint32_t GetConnectionWindowUs() const { return mIntervalUs * kConnectionRetryInterval * kConnectionRetryCount; }
+
+    /**
+     * Returns the end of the wake-up sequence time.
+     *
+     * @returns End of the wake-up sequence time.
+     */
+    TimeMicro GetTxEndTime() const { return mTxEndTimeUs; }
+
+    /**
+     * Stops the ongoing wake-up sequence.
+     */
+    void Stop(void);
+
+private:
+    using WakeupTimer = TimerMicroIn<Mac::Mac, &Mac::Mac::RequestWakeupFrameTransmission>;
+
+    constexpr static uint8_t kConnectionRetryInterval = OPENTHREAD_CONFIG_WAKEUP_COORDINATOR_CONNECTION_RETRY_INTERVAL;
+    constexpr static uint8_t kConnectionRetryCount    = OPENTHREAD_CONFIG_WAKEUP_COORDINATOR_CONNECTION_RETRY_COUNT;
+
+    /*
+     * Called by the MAC layer when a wake-up frame transmission is about to be started.
+     */
+    Mac::TxFrame *PrepareWakeupFrame(Mac::TxFrames &aTxFrames);
+
+    /*
+     * Called by the MAC layer when a wake-up frame transmission is done.
+     */
+    void ScheduleNext(bool aIsFirstFrame = false);
+
+    Mac::ExtAddress mWedAddress;
+    TimeMicro       mTxTimeUs;             //< Point in time when the next TX occurs.
+    TimeMicro       mTxEndTimeUs;          //< Point in time when the wake-up sequence is over.
+    const uint16_t  mTxRequestAheadTimeUs; //< How much ahead the TX MAC operation needs to be requested.
+    uint16_t        mIntervalUs;           //< Interval between consecutive wake-up frames.
```

Comment:
Can replace `///<` with `//` (since we don't want doxygen docs to be generated for `private` definition).